idseq_pipeline/commands/assembly_functions.py

```diff
@@ -0,0 +1,134 @@
+TEMP_DIR = ROOT_DIR + '/tmp' # tmp directory with a lot of space for sorting large files
+SPADES_DIR = ROOT_DIR + '/spades' # outputs of SPAdes assemblies go here
+
+# parameters
+ASSEMBLY_READ_THRESHOLD = 100
+
+# arguments from environment variables
+ALIGNMENT_S3_PATH = os.environ.get('ALIGNMENT_S3_PATH').rstrip('/')
+POSTPROCESS_S3_PATH = os.environ.get('POSTPROCESS_S3_PATH').rstrip('/')
+AWS_BATCH_JOB_ID = os.environ.get('AWS_BATCH_JOB_ID', 'local')
+sample_name = ALIGNMENT_S3_PATH[5:].rstrip('/').replace('/', '-')
+SAMPLE_DIR = DEST_DIR + '/' + sample_name
+INPUT_DIR = SAMPLE_DIR + '/inputs'
+RESULT_DIR = SAMPLE_DIR + '/results'
+
+# input files
+TAXON_COUNTS = 'multihit_idseq_web_sample.json'
+TAXID_ANNOT_FASTA = 'taxid_annot.fasta'
+
+# outputs
+SAMPLE_S3_OUTPUT_PATH = POSTPROCESS_S3_PATH
+ASSEMBLY_DIR = 'assembly'
+STATUS_FILE = 'job-complete'
+
+def run_stage4():
+
+    # make data directories
+    execute_command("mkdir -p %s %s %s %s %s" % (SAMPLE_DIR, RESULT_DIR, REF_DIR, TEMP_DIR, SPADES_DIR))
+
+    # download input
+    execute_command("aws s3 cp --quiet %s/%s %s/" % (ALIGNMENT_S3_PATH, TAXON_COUNTS, INPUT_DIR))
+    pipeline_output_json = os.path.join(INPUT_DIR, TAXON_COUNTS)
+    execute_command("aws s3 cp --quiet %s/%s %s/" % (POSTPROCESS_S3_PATH, TAXID_ANNOT_FASTA, INPUT_DIR))
+    full_fasta = os.path.join(INPUT_DIR, TAXID_ANNOT_FASTA)
+
+    # run assembly
+    def get_taxids_to_assemble(taxon_counts):
+        total_counts = {} # sum of NT and NR counts for each taxid
+        filtered_total_counts = {} # taxids that satisfy count threshold
+        for item in taxon_counts:
+            taxid = item['tax_id']
+            total_counts[taxid] = total_counts.get(taxid, 0) + item['count']
+            if total_counts[taxid] >= ASSEMBLY_READ_THRESHOLD:
+                filtered_total_counts[taxid] = total_counts[taxid]
+        sorted_filtered_taxids = sorted(filtered_total_counts, key=filtered_total_counts.get, reverse=True)        
+        return sorted_filtered_taxids
+
+    def make_inputs_for_assembly():
+        # Get taxids to assemble based on criteria from report, currently: taxids with read count >= ASSEMBLY_READ_THRESHOLD
+        with open(pipeline_output_json) as f:
+            pipeline_output = json.load(f)
+        taxon_counts = pipeline_output['pipeline_output']['taxon_counts_attributes']
+        sorted_taxids_to_assemble = get_taxids_to_assemble(taxon_counts)
+        # Get reads for those taxids
+        output = {}
+        hit_delimiters = ['_nt', '_nr'] # annotations are like e.g. "genus_nt:543:"
+        for taxid in sorted_taxids_to_assemble:
+            pattern = '\|'.join(['%s:%s:' % (delimiter, taxid) for delimiter in hit_delimiters])
+            partial_fasta =  os.path.join(RESULT_DIR, taxid + ".fasta")
+            try:
+                subprocess.check_call("grep -A 1 --no-group-separator '%s' %s > %s" % (pattern, full_fasta, partial_fasta), shell=True)
```

Comment:
For efficiency you could build a single grep regular expression, and run grep once.   Not sure how much it matters for this here.  How long does this take for runs with 10k or 20k taxids?